File: Easy/RansomNote.java
//https://leetcode.com/problems/ransom-note/

class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        if (ransomNote.length() > magazine.length()) {
            return false;
        }

        Map<Character, Integer> map = new HashMap<>();

        for (int i = 0; i < magazine.length(); i++) {
            map.put(magazine.charAt(i), map.getOrDefault(magazine.charAt(i), 0) + 1);
        }

        for (int j = 0; j < ransomNote.length(); j++) {
            map.put(ransomNote.charAt(j), map.getOrDefault(ransomNote.charAt(j), 0) - 1);
        }

        for (Map.Entry<Character, Integer> entry : map.entrySet()) {
            if (entry.getValue() < 0) {
                return false;
            }
        }

        return true;

    }
}

File: Easy/RansomNote/RansomNote.py
#https://leetcode.com/problems/ransom-note/

# class Solution:
#     def canConstruct(self, ransomNote: str, magazine: str) -> bool:
#         if len(magazine) < len(ransomNote):
#             return False

#         letters = [0]*26

#         for i in range(0, len(magazine)):
#             letters[ord(magazine[i]) - ord('a')] = letters[ord(magazine[i]) - ord('a')]+1

#         for i in range(0,len(ransomNote)):
#             if letters[ord(ransomNote[i]) - ord('a')] == 0:
#                 return False
#             else:
#                 letters[ord(ransomNote[i]) - ord('a')] = letters[ord(ransomNote[i]) - ord('a')]-1

#         return True

class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        return Counter(ransomNote) <= Counter(magazine)

# Counter(ransomNote) creates a dictionary where the keys are the characters in ransomNote, and the values are the counts of each character.

# Similarly, Counter(magazine) creates a dictionary with the counts of each character in magazine.

# The comparison Counter(ransomNote) <= Counter(magazine) checks if all the counts of characters in ransomNote are less than or equal to the counts of the corresponding characters in magazine. If this condition is true, it means that there are enough characters in magazine to construct ransomNote.

# The method returns True if the condition is met (meaning ransomNote can be constructed from magazine), and False otherwise.